system-addon/lib/TopSitesFeed.jsm

```diff
@@ -210,23 +246,41 @@ this.TopSitesFeed = class TopSitesFeed {
 
   /**
    * Pin a site at a specific position saving only the desired keys.
+   * @param customScreenshotURL {string} User set URL of preview image for site
+   * @param label {string} User set string of custom site name
    */
-  _pinSiteAt({label, url}, index) {
+  async _pinSiteAt({customScreenshotURL, label, url}, index) {
     const toPin = {url};
     if (label) {
       toPin.label = label;
     }
+    if (customScreenshotURL) {
+      toPin.customScreenshotURL = customScreenshotURL;
+    }
     NewTabUtils.pinnedLinks.pin(toPin, index);
+
+    await this._updateLinkCustomScreenshot({customScreenshotURL, url});
+  }
+
+  async _updateLinkCustomScreenshot(site) {
+    // If screenshot url changed or was removed we need to update the cached link obj
```

Comment:
This comment says "update" but more accurately it's clearing?

Reply:
Renamed to `_clearLinkCustomScreenshot`